framework/src/meshmodifiers/BreakBoundaryOnSubdomain.C

```diff
@@ -0,0 +1,106 @@
+  params.addParam<std::vector<BoundaryName> >("boundaries", "Boundaries to be broken. Default means to break all boundaries");
+  return params;
+}
+
+BreakBoundaryOnSubdomain::BreakBoundaryOnSubdomain(const InputParameters & parameters) :
+    MeshModifier(parameters)
+{
+}
+
+void
+BreakBoundaryOnSubdomain::modify()
+{
+  // get the mesh and boundary info
+  auto & mesh = _mesh_ptr->getMesh();
+  auto & boundary_info = mesh.get_boundary_info();
+
+  // get IDs of all boundaries to be broken
+  std::set<BoundaryID> breaking_boundary_ids;
+  if (isParamValid("boundaries"))
+  {
+    auto & boundary_names = getParam<std::vector<BoundaryName> >("boundaries");
+    for (auto i = beginIndex(boundary_names); i < boundary_names.size(); ++i)
+      breaking_boundary_ids.insert(_mesh_ptr->getBoundaryID(boundary_names[i]));
+  }
+  else
+    breaking_boundary_ids = boundary_info.get_boundary_ids();
+
+  auto end_el = mesh.active_elements_end();
+
+  // create a list of new boundary names
+  std::set<BoundaryName> new_boundary_name_set;
+  for (auto el = mesh.active_elements_begin(); el != end_el; ++el)
+  {
+    auto elem = *el;
+    auto subdomain_id = elem->subdomain_id();
+    auto subdomain_name = mesh.subdomain_name(subdomain_id);
+    if (subdomain_name == "")
+      subdomain_name = std::to_string(subdomain_id);
+    for (unsigned int side = 0; side < elem->n_sides(); ++side)
+    {
+      auto side_boundary_ids = boundary_info.boundary_ids(elem, side);
+      for (auto i = beginIndex(side_boundary_ids); i < side_boundary_ids.size(); ++i)
+        if (breaking_boundary_ids.count(side_boundary_ids[i]) > 0)
+          new_boundary_name_set.insert(boundary_info.sideset_name(side_boundary_ids[i]) + "_to_" + subdomain_name);
+    }
+  }
+
+  // assign boundary IDs to the boundaries to be added
+  std::vector<BoundaryName> new_boundary_names;
+  std::copy(new_boundary_name_set.begin(), new_boundary_name_set.end(), std::back_inserter(new_boundary_names));
+  auto new_boundary_ids = _mesh_ptr->getBoundaryIDs(new_boundary_names, true);
+
+  // assign boundary names to the new boundaries
+  for (auto i = beginIndex(new_boundary_ids); i < new_boundary_ids.size(); ++i)
+  {
+    boundary_info.sideset_name(new_boundary_ids[i]) = new_boundary_names[i];
+    boundary_info.nodeset_name(new_boundary_ids[i]) = new_boundary_names[i];
+  }
+
+  // add sides into the side sets
+  for (auto el = mesh.active_elements_begin(); el != end_el; ++el)
+  {
+    auto elem = *el;
+    auto subdomain_id = elem->subdomain_id();
+    auto subdomain_name = mesh.subdomain_name(subdomain_id);
+    if (subdomain_name == "")
+      subdomain_name = std::to_string(subdomain_id);
+    for (unsigned int side = 0; side < elem->n_sides(); ++side)
+    {
+      auto side_boundary_ids = boundary_info.boundary_ids(elem, side);
+      for (auto i = beginIndex(side_boundary_ids); i < side_boundary_ids.size(); ++i)
+      {
+        if (breaking_boundary_ids.count(side_boundary_ids[i]) > 0)
```

Comment:
Oh, I misunderstood this line of code! Actually this is really odd. I'm not even sure why `count` exists on a set because it is always zero or one. `std::set::find()` would definitely make more sense here.